### PR TITLE
Reuse code between CS direct and indirect variables

### PR DESCRIPTION
**Pull request by triceo**
This is a pure refactor. `DirectPatternVariable` and `IndirectPatternVariable` were nearly carbon copies. I extracted the common logic into `AbstractPatternVariable`, significantly reducing code duplication.

### CODE REVIEW

File: core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/drools/common/AbstractPatternVariable.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+     * This method abstracts this behavior, so that the surrounding code may be shared between both implementations.
+     *
+     * @param patternVar never null, pattern variable to extract the value from
+     * @return value of the variable
+     */
+    protected abstract A extract(PatternVar_ patternVar);
+
+    protected abstract Child_ create(UnaryOperator<PatternDSL.PatternDef<PatternVar_>> patternMutator);
+
+    protected abstract Child_ create(ViewItem<?> dependentExpression);
```
Comment (triceo) [inline core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/drools/common/AbstractPatternVariable.java:123]: The extending classes are now required to implement this.
All the other code in this class has been moved from the extending classes, without modification.

**Comment by kie-ci1** [comment]
**(tests) - optaplanner job** `#364` was: **FAILURE**
Possible explanation: Pipeline failure or project build failure

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/optaplanner.tests.optaplanner/364/display/redirect or see console log:
<details>
<summary><b>Console Logs</b></summary>
[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list__item_icon_MarginRight.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list__item_icon_MarginTop.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list__item_icon_MinWidth.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_li_MarginTop.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_m_bordered_li_BorderBottomColor.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_m_bordered_li_BorderBottomWidth.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_m_bordered_li_PaddingBottom.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_m_icon_lg__item_icon_FontSize.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_m_icon_lg__item_icon_MarginRight.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-employee-rostering_ws-cleanup_1645532483592/bc/kiegroup_optaweb_employee_rostering/optaweb-employee-rostering/optaweb-employee-rostering-frontend/node_modules/@patternfly/react-tokens/dist/js/c_list_m_icon_lg__item_icon_MinWidth.js’: No such file or directory<br/>[2022-02-22T12:21:28.725Z] 780M	/home/jenkins/workspace/KIE/kogito/main<br/>[2022-02-22T12:21:28.725Z] 780M	/home/jenkins/workspace/KIE/kogito<br/>[2022-02-22T12:21:28.725Z] 780M	/home/jenkins/workspace/KIE<br/>[2022-02-22T12:21:28.725Z] 780M	/home/jenkins/workspace<br/>[Pipeline] }<br/>[Pipeline] // dir<br/>[Pipeline] }<br/>[Pipeline] // script<br/>[Pipeline] }<br/>[Pipeline] // withEnv<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Install build-chain tool)<br/>Stage "Install build-chain tool" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Build projects)<br/>Stage "Build projects" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Sonar analysis)<br/>Stage "Sonar analysis" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (check space after build)<br/>Stage "check space after build" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Declarative: Post Actions)<br/>[Pipeline] script<br/>[Pipeline] {<br/>[Pipeline] sh<br/>[2022-02-22T12:21:29.593Z] + wget --no-check-certificate -qO - https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/optaplanner.tests.optaplanner/364/api/json<br/>[Pipeline] readJSON<br/>[Pipeline] sh<br/>
</details>

**Comment by kie-ci1** [comment]
**(tests) - optaweb-vehicle-routing job** `#365` was: **FAILURE**
Possible explanation: Pipeline failure or project build failure

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/optaplanner.tests.downstream.optaweb-vehicle-routing/365/display/redirect or see console log:
<details>
<summary><b>Console Logs</b></summary>
[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/cheese-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-bishop-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-board-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-king-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-knight-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-pawn-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] du: cannot access ‘/home/jenkins/workspace/KIE/kogito/main/pullrequest/optaplanner.tests.downstream.optaweb-vehicle-routing_ws-cleanup_1645532537628/bc/kiegroup_optaweb_vehicle_routing/optaweb-vehicle-routing/optaweb-vehicle-routing-frontend/node_modules/@patternfly/react-icons/dist/js/icons/chess-queen-icon.d.ts’: No such file or directory<br/>[2022-02-22T12:22:29.610Z] 673M	/home/jenkins/workspace/KIE/kogito/main<br/>[2022-02-22T12:22:29.610Z] 673M	/home/jenkins/workspace/KIE/kogito<br/>[2022-02-22T12:22:29.610Z] 0	/home/jenkins/workspace/KIE/7.x/compile<br/>[2022-02-22T12:22:29.610Z] 0	/home/jenkins/workspace/KIE/7.x<br/>[2022-02-22T12:22:29.610Z] 673M	/home/jenkins/workspace/KIE<br/>[2022-02-22T12:22:29.610Z] 673M	/home/jenkins/workspace<br/>[Pipeline] }<br/>[Pipeline] // dir<br/>[Pipeline] }<br/>[Pipeline] // script<br/>[Pipeline] }<br/>[Pipeline] // withEnv<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Install build-chain tool)<br/>Stage "Install build-chain tool" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Build projects)<br/>Stage "Build projects" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Sonar analysis)<br/>Stage "Sonar analysis" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (check space after build)<br/>Stage "check space after build" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Declarative: Post Actions)<br/>[Pipeline] script<br/>[Pipeline] {<br/>[Pipeline] sh<br/>[2022-02-22T12:22:30.587Z] + wget --no-check-certificate -qO - https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/optaplanner.tests.downstream.optaweb-vehicle-routing/365/api/json<br/>[Pipeline] readJSON<br/>[Pipeline] sh<br/>
</details>

**Review by rsynek — APPROVED** [review]
LGTM

**Pull request merged**